upstairs/src/lib.rs

```diff
@@ -192,6 +192,7 @@ pub struct CrucibleOpts {
     pub key_pem: Option<String>,
     pub root_cert_pem: Option<String>,
     pub control: Option<SocketAddr>,
+    pub read_only: bool,
```

Comment:
Just a thought to put in the back of your head here... What would we do if we wanted to support a RW upstairs changing it's mind at some point and wanting to go RO?

I was thinking about the situation where there was some failure, and we could no longer do writes,  could the upstairs switch to read only?   There would be a bunch of other stuff.. but.. just thinking..

Reply:
You know, I think we could support something like this. If we send back a `CrucibleError` up to Propolis, it'll bubble up through the device emulation as an IO error of some sort. Some quick searching shows that [devices do set themselves to read-only when they see errors](https://www.reddit.com/r/techsupport/comments/hx3qkw/nvme_ssds_suddenly_marked_as_readonly/). We could set something similar in the device emulation, matching on a certain error variant.

I think there's situations where we would want to switch an Upstairs to read-only. For example, one that comes to mind is if there's only one downstairs that's healthy and active. We can't accept writes because they won't hit at least two downstairs.

It's a good idea :)

Reply:
> It's a good idea :)

Crazy that all those people in the past came to the future and stole my original idea then went back to the past and implemented it.